packages/browser/src/plugins/in-app-plugin/index.ts

```diff
@@ -0,0 +1,149 @@
+    siteId: string | undefined
+    events: EventListenerOrEventListenerObject | null | undefined
+
+    _env: string | undefined
+    _logging: boolean | undefined
+}
+
+export function InAppPlugin(
+    settings: InAppPluginSettings,
+  ): Plugin {
+
+    let _analytics: Analytics;
+    let _gistLoaded:boolean = false;
+    let _pluginLoaded:boolean = false;
+    const _eventTarget:EventTarget = new EventTarget();
+
+    function attachListeners() {
+        if(!_gistLoaded || _pluginLoaded)
+            return;
+
+        _analytics.on('reset', reset);
+
+        if(settings.events) {
+            allEvents.forEach((event) => {
+                _eventTarget.addEventListener(event, settings?.events as EventListenerOrEventListenerObject);
+            });
+            ['messageShown', 'messageDismissed', 'messageError'].forEach((event) => {
+                Gist.events.on(event, (message: any) => {
+                    _eventTarget.dispatchEvent(newEvent(gistToCIO(event), { 
+                        messageId: message.messageId,
+                        deliveryId: message.properties?.gist?.campaignId,
+                    }));
+                });
+            });
+        }
+
+        Gist.events.on('messageShown', (message: any) => {
+            const deliveryId:string = message?.properties?.gist?.campaignId;
+            if (typeof deliveryId != 'undefined' && deliveryId != '') {
+                _analytics.track(semanticEvents.JourneyMetric, {
+                    'deliveryId': deliveryId,
+                    'metric': semanticEvents.Opened,
+                });
+            }
+        });
+
+        Gist.events.on('messageAction', (params: any) => {
+            const deliveryId:string = params?.message?.properties?.gist?.campaignId;
+            if (params.action != 'gist://close' && typeof deliveryId != 'undefined' && deliveryId != '') {
+                _analytics.track(semanticEvents.JourneyMetric, {
+                    'deliveryId': deliveryId,
+                    'metric': semanticEvents.Clicked,
+                    'actionName': params.name,
+                    'actionValue': params.action,
+                });
+            }
+            settings.events && _eventTarget.dispatchEvent(newEvent(InAppEvents.MessageAction, {
+                messageId: params.message.messageId,
+                deliveryId: deliveryId,
+                action: params.action,
+                name: params.name,
+                actionName: params.name,
+                actionValue: params.action,
+                message:{
+                    dismiss: function() {
+                        Gist.dismissMessage(params?.message?.instanceId);
+                    }
+                }
+             }));
+        });
+    }
+
+    async function page(ctx: Context): Promise<Context> {
+        if(!_pluginLoaded)
+            return ctx;
+
+        let page:string = ctx.event?.properties?.name || ctx.event?.properties?.url;
+        if(page != "") {
```

Comment:
This will be true if page is `undefined` for example. Is that expected?

Reply:
Nope, I'll do the same check as user below: `if (typeof page === 'string' && page.length > 0)`